File: cplusplus/TestBST.cpp
#include <algorithm>
#include <iostream>
#include <vector>

#include "BST.hpp"

using std::find;
using std::sort;
using std::size_t;
using std::cout;
using std::endl;
using std::vector;

template <typename T>
static std::ostream& operator<<(std::ostream &out, const vector<T>& v) {
    out << "[";
    auto itr = v.cbegin();
    const auto &end = v.cend();
    if (itr != end) {
        out << *itr++;
    }
    while (itr != end) {
        out << ", " << *itr++;
    }
    out << "]";
    return out;
}

template <typename U , typename V>
static bool verify(U expected, V actual) {
    if (expected != actual) {
        cout << "expected " << expected << ", received " << actual << endl;
        return false;
    }
    return true;
}

static bool verify(int expected, size_t actual) {
    return verify(static_cast<size_t>(expected), actual);
}

int main(int, char **) {
    bool status = true;
    BST::BST<int> b;
    vector<int> exp;
    status |= verify(0, b.size());
    status |= verify(exp, b.list());
    b.add(5);
    exp.push_back(5);
    status |= verify(true, b.contains(5));
    status |= verify(1, b.size());
    status |= verify(exp, b.list());
    for (int i = 10; i >= 0; --i) {
        b.add(i);
        if (find(exp.cbegin(), exp.cend(), i) == exp.cend()) {
            exp.push_back(i);
        }
    }
    status |= verify(11, b.size());
    sort(exp.begin(), exp.end());
    status |= verify(exp, b.list());
    for (int i = 0; i < 11; ++i) {
        status |= verify(true, b.contains(i));
    }
    status |= verify(false, b.contains(11));
    return !status;
}

File: prolog/test_BST.pl
%% -*- Mode: Prolog; -*-

comp(A, A, 0).
comp(A, B, -1) :- A < B.
comp(A, B, 1) :- A > B.

verify(A, A).
verify(Exp, Act) :- Exp \= Act,
                    format('expected ~p, received ~p~n', [Exp, Act]).

addEach(_, T, T, A, Hi) :- A =:= Hi + 1.
addEach(Comp, T, T1, Itr, Hi) :- Itr =< Hi,
                                 add(T, Itr, Comp, TR),
                                 I is Itr + 1,
                                 addEach(Comp, TR, T1, I, Hi).

containsEach(_, A, Hi) :- A =:= Hi + 1.
containsEach(T, Itr, Hi) :- Itr =< Hi,
                            contains(T, Itr),
                            I is Itr + 1,
                            containsEach(T, I, Hi).

go :-
    T1 = empty,
    size(T1, S1), verify(0, S1),
    list(T1, L1), verify([], L1),
    add(T1, 5, comp, T2),
    contains(T2, 5),
    size(T2, S2), verify(1, S2),
    list(T2, L2), verify([5], L2),
    addEach(comp, T2, T3, 0, 10),
    size(T3, S3), verify(11, S3),
    list(T3, L3), verify([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10], L3),
    containsEach(T3, 0, 10),
    \+ contains(T3, 11).
